### PR TITLE
[OCPCLOUD-1502] RollingUpdate Strategy

### PR DESCRIPTION
**Pull request by damdo**
This introduces an implementation of the `RollingUpdate` Strategy for the ControlPlaneMachineSet Machines.

### CODE REVIEW

File: pkg/controllers/controlplanemachineset/updates.go [hunk 1]
```diff
@@ -137,6 +139,85 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineUpdates(ctx context.C
...
+			// Wait for it to become Ready.
+			logger = logger.WithValues("index", idx, "namespace", r.Namespace, "name", getMachineReplacementName(idx))
+			logger.V(2).Info(waitingForReady)
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// Reconcile machines that need an update.
+	for idx, machines := range sortedIndexedMs {
+		// Find out if and what Machines in this index need an update.
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:180]: Might be worth breaking the content of this loop into a helper function

File: pkg/controllers/controlplanemachineset/updates.go [hunk 2]
```diff
@@ -137,6 +139,85 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineUpdates(ctx context.C
...
+	// Indexes are sorted in ascendent order, so that all the operations of the same importance,
+	// are executed prioritizing the lower indexes first.
+	sortedIndexedMs := sortMachineInfos(indexedMachineInfos)
+
+	// The maximum number of machines that
+	// can be scheduled above the original number of desired machines.
+	// At present, the surge is limited to a single Machine instance.
+	maxSurge := 1
+	// Devise the existing surge and keep track of the current surge count.
+	surgeCount := deviseExistingSurge(cpms, sortedIndexedMs)
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:155]: What happens if the surgeCount is already met? Should there be an early exit here or is this handled later? 

I think the answer is no because we need to handle deletes, but perhaps there should be a comment to stop people like me asking why we don't do it here

File: pkg/controllers/controlplanemachineset/updates.go [hunk 3]
```diff
@@ -152,3 +233,134 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineRollingUpdate(ctx con
...
+
+// getMachineName returns the canonical name for a machine of a certain index.
+func getMachineName(idx int) string {
+	return "machine-" + strconv.Itoa(idx)
+}
+
+// getMachineReplacementName returns the canonical name for a replacement machine of a certain index.
+func getMachineReplacementName(idx int) string {
+	return "machine-replacement-" + strconv.Itoa(idx)
+}
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:296]: Didn't expect to see these here, where are they being used? I would expect to be getting names from the MachineInfo not from a function like this, the names will vary per machine right?

File: pkg/controllers/controlplanemachineset/updates.go [hunk 4]
```diff
@@ -152,3 +233,134 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineRollingUpdate(ctx con
...
+
+// sortMachineInfos returns a sorted list of indexes together with their MachineInfos.
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:298]: Sort by what?

File: pkg/controllers/controlplanemachineset/updates.go [hunk 5]
```diff
@@ -152,3 +233,134 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineRollingUpdate(ctx con
...
 	return ctrl.Result{}, nil
 }
+
+// isDeletedMachine checks if a machine is deleted.
+func isDeletedMachine(m machineproviders.MachineInfo) bool {
+	return m.MachineRef.ObjectMeta.DeletionTimestamp != nil
+}
+
+// needUpdateMachines returns the list of MachineInfo which have Machines that need an update.
+func needUpdateMachines(machinesInfo []machineproviders.MachineInfo) []machineproviders.MachineInfo {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:243]: There's a lot of repeated looping here, what if you had a `filterMachines` function that took filters as an argument that could then be composed, would this make it clearer or more complicated?

```
type filterMachines func(machineprovider.MachineInfo) bool

func filterMachines(machinesInfo []machineproviders.MachineInfo, filters ...filterFunc) []machineproviders.MachineInfo {
  filtered := []machineproviders.MachineInfo{}

  for _, m := range machineInfos {
    for _, filter := range filters {
      if !filter(m) {
        continue
      }
    }

    filtered = append(filtered, m)
  }

  return filtered
}
```

Then each of your other helpers can become a combination of filters into this 🤔 

```
func not(filter filterFunc) filterFunc {
  return func(m machineproviders.MachineInfo) bool {
    return !filter(m)
  }
}

func machineNeedsUpdate(m machineproviders.MachineInfo) bool {
  return m.NeedsUpdate
}

func machineIsReady(m machineproviders.MachineInfo) bool {
  return m.Ready
}

func pendingMachines(machinesInfo []machineproviders.MachineInfo) []machineproviders.MachineInfo {
  return filterMachines(machinesInfo, not(machineIsReady), not(machineNeedsUpdate))
}

func updatedMachines(machinesInfo []machineproviders.MachineInfo) []machineproviders.MachineInfo {
  return filterMachines(machinesInfo, machineIsReady, not(machineNeedsUpdate))
}

func readyMachines(machinesInfo []machineproviders.MachineInfo) []machineproviders.MachineInfo {
  return filterMachines(machinesInfo, machineIsReady)
}

```

File: pkg/controllers/controlplanemachineset/updates.go [hunk 6]
```diff
@@ -137,6 +139,85 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineUpdates(ctx context.C
...
+			if any(pendingMachines(machines)) {
+				// A Pending Machine Replacement is present.
+				// Wait for it to become Ready.
+				logger.V(2).WithValues("replacementName", getMachineReplacementName(idx)).Info(waitingForReplacement)
+				return ctrl.Result{}, nil
+			}
+
+			// No Healthy or Pending Replacement Machine exists,
+			// trigger a Machine creation.
+			return createMachine(ctx, logger, machineProvider, idx, maxSurge, &surgeCount)
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:214]: Handling the surge inside another function means that, in the future, if we need to create multiple machines, we won't be able to right? Is there a way we can keep track of the surge within this function itself, a short circuit at the top of the loop maybe?

Or do we not do this because of deletes? Perhaps we should handle this in two phases? A delete phase, then an create phase? Or vice versa?

Given the cap of Machines in the control plane should be about 5 instances, I'm not concerned about the cost of iterating over a loop a few times if it makes the code clearer

File: pkg/controllers/controlplanemachineset/updates.go [hunk 7]
```diff
@@ -137,6 +139,85 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineUpdates(ctx context.C
...
+	}
+
+	// Reconcile any index with no Ready Machines but a replacement pending.
+	for idx, machines := range sortedIndexedMs {
+		if empty(readyMachines(machines)) && any(pendingMachines(machines)) {
+			// There are No Ready Machines for this index but a Pending Machine Replacement is present.
+			// Wait for it to become Ready.
+			logger = logger.WithValues("index", idx, "namespace", r.Namespace, "name", getMachineReplacementName(idx))
+			logger.V(2).Info(waitingForReady)
+			return ctrl.Result{}, nil
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:174]: Should we process all indexes here? Rather than just one?

File: pkg/controllers/controlplanemachineset/updates.go (hunk 2 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go:155]: Added some extra explanation 👍

File: pkg/controllers/controlplanemachineset/updates.go (hunk 3 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go]: Oops, this is a refuse. Yes just pushed the right one.

File: pkg/controllers/controlplanemachineset/updates.go (hunk 5 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go:251]: Hhm yeah I can see the functional side of me being tempted by this.
Although it would be purely cosmetic and adds quite a bit of boilerplate to avoid looping over a list a couple of times.

Personally I would keep 1/2 row `for` repetitions as I wouldn't expect loops changing that much and needing a lot of refactoring to justify the extra complexity. But I do not feel strongly about this.

File: pkg/controllers/controlplanemachineset/updates.go (hunk 6 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go:222]: > Handling the surge inside another function means that, in the future, if we need to create multiple machines, we won't be able to right?

If in the future we were to increase `maxSurge` to let's say `2`, the current logic would still allow creating multiple machines up to when we hit the surge. It won't allow creating multiple in the same reconciliation though and this is a conscious design choice I made in the code after experimenting things a bit.

Writing the logic in a way that accounts multiple indexes involved at every loop, would complicate the logic significantly.

Unless there already are assumptions on how we do this, I'd prefer to limit reconciliation to a single index per reconcile. This allows us to perform operations in a per index incremental lock step fashion and to keep complexity fairly low.

File: pkg/controllers/controlplanemachineset/updates.go (hunk 7 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go:179]: It would break tests as they are written now.
In this case, for example, tests expect just a single "waitingForReady" log before the return.

So for the motivations described here https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/56#discussion_r913734444 I'd limit it to one index at a time.

File: pkg/controllers/controlplanemachineset/updates.go [hunk 8]
```diff
@@ -137,6 +138,97 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineUpdates(ctx context.C
...
+
+	// Reconcile any index with no Machine first.
+	for idx, machines := range sortedIndexedMs {
+		if isEmpty(machines) {
+			// There are No Machines for this index.
+			// Create a new Machine for it.
+			logger = logger.WithValues("index", idx, "namespace", r.Namespace, "name", "<Unknown>")
+			return createMachine(ctx, logger, machineProvider, idx, maxSurge, &surgeCount)
+		}
+	}
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:166]: What happens if there's no capacity to surge and an index is empty? 

Eg Index 1 has 2 instances - 1 old, 1 updated, index 0 has 2 instances, 1 old, 1 updated, index 2 has no instances, then we will never get to deleting Machines, instead we will loop on this and fail to create forever

This goes back to, I think we need to do this in multiple phases, do a delete phase, then a create phase

File: pkg/controllers/controlplanemachineset/updates.go [hunk 9]
```diff
@@ -137,6 +138,97 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineUpdates(ctx context.C
...
+
+	// Reconcile any index with no Ready Machines but a replacement pending.
+	for idx, machines := range sortedIndexedMs {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:169]: Likewise here I think we can do the delete phase first

File: pkg/controllers/controlplanemachineset/updates.go [hunk 10]
```diff
@@ -152,3 +241,124 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineRollingUpdate(ctx con
...
+	return currentReplicas - desiredReplicas
+}
+
+// hasAny checks if a MachineInfo slice contains at least 1 element.
+func hasAny(machinesInfo []machineproviders.MachineInfo) bool {
+	return len(machinesInfo) > 0
+}
+
+// any checks if a MachineInfo slice is empty.
+func empty(machinesInfo []machineproviders.MachineInfo) bool {
```
[the remark this answers is not in the document]
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go]: Yes that's in a followup commit already :)

File: pkg/controllers/controlplanemachineset/updates.go (hunk 9 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go]: Agree, delete phase is now first.

File: pkg/controllers/controlplanemachineset/updates.go [hunk 11]
```diff
@@ -152,3 +241,124 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineRollingUpdate(ctx con
...
+	for i := range machinesInfo {
+		if machinesInfo[i].Ready {
+			result = append(result, machinesInfo[i])
+		}
+	}
+	return result
+}
+
+// sortMachineInfos returns a list numerically sorted by index, of each index' MachineInfos.
+func sortMachineInfos(indexedMachineInfos map[int32][]machineproviders.MachineInfo) [][]machineproviders.MachineInfo {
```
[the remark this answers is not in the document]
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go]: Definitely! Added.

File: pkg/controllers/controlplanemachineset/updates.go [hunk 12]
```diff
@@ -152,3 +202,243 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineRollingUpdate(ctx con
...
+			logger.V(2).Info(waitingForReady)
+
+			return true, ctrl.Result{}, nil
+		}
+	}
+
+	return false, ctrl.Result{}, nil
+}
+
+func (r *ControlPlaneMachineSetReconciler) reconcileIndexWithOutdatedButUpdated(ctx context.Context, logger logr.Logger, sortedIndexedMs [][]machineproviders.MachineInfo, machineProvider machineproviders.MachineProvider) (bool, ctrl.Result, error) {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:361]: The name of this function, makes it hard to know what it is actually doing IMO.

I would expect this to be something more like, `deleteReplacedMachines` or something similar, as that is what we are doing, we are looking for any machine that has successfully been replaced, and deleting it

File: pkg/controllers/controlplanemachineset/updates.go [hunk 13]
```diff
@@ -152,3 +202,243 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineRollingUpdate(ctx con
...
+		return ctrl.Result{}, werr
+	}
+
+	logger.V(2).Info(createdReplacement)
+	*surgeCount++
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ControlPlaneMachineSetReconciler) reconcileIndexNonReadyButPending(ctx context.Context, logger logr.Logger, sortedIndexedMs [][]machineproviders.MachineInfo) (bool, ctrl.Result, error) {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:342]: The name of this doesn't make it obvious to me what it's doing, I would expect it to be more inline with what it's achieving, so maybe `waitForPendingIndex` or something like that? We are waiting for an index that has no ready machines but has a pending machine right?

File: pkg/controllers/controlplanemachineset/updates.go [hunk 14]
```diff
@@ -152,3 +202,243 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineRollingUpdate(ctx con
...
+			logger.V(2).Info(waitingForRemoved)
+
+			return true, ctrl.Result{}, nil
+		}
+	}
+
+	return false, ctrl.Result{}, nil
+}
+
+func (r *ControlPlaneMachineSetReconciler) reconcileIndexWithOutdatedButPending(ctx context.Context, logger logr.Logger, sortedIndexedMs [][]machineproviders.MachineInfo, machineProvider machineproviders.MachineProvider) (bool, ctrl.Result, error) {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:387]: As with the others, lets make the name say what this is doing, `waitForPendingMachines`? Why do we do this separately from the `non ready` index, perhaps they can be combined?

File: pkg/controllers/controlplanemachineset/updates.go [hunk 15]
```diff
@@ -152,3 +202,243 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineRollingUpdate(ctx con
...
+			logger := logger.WithValues("index", int(outdatedMachine.Index), "namespace", r.Namespace, "name", outdatedMachine.MachineRef.ObjectMeta.Name)
+
+			result, err := createMachine(ctx, logger, machineProvider, outdatedMachine.Index, maxSurge, surgeCount)
+
+			return true, result, err
+		}
+	}
+
+	return false, ctrl.Result{}, nil
+}
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:444]: Perhaps we want to combine these to a `createReplacementMachines` which has two loops, first a not about updating indexes with no machines, then the second loop for those that have a Machine needing an update

File: pkg/controllers/controlplanemachineset/updates.go [hunk 16]
```diff
@@ -137,6 +142,51 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineUpdates(ctx context.C
...
+
+	// Reconcile any index with No Machines.
+	if done, result, err := r.reconcileIndexWithNoMachine(ctx, logger, sortedIndexedMs, machineProvider, maxSurge, &surgeCount); err != nil || done {
+		return result, err
+	}
+
+	// Reconcile any index with any remaining Outdated Machine.
+	if done, result, err := r.reconcileIndexWithOutdated(ctx, logger, sortedIndexedMs, machineProvider, maxSurge, &surgeCount); err != nil || done {
+		return result, err
+	}
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:185]: Because we are using the terms of the classes of Machines, and not really talking about the actions that we are taking, I'm finding the logic of this function hard to follow.

I would have expected something more like a story

```
// Delete machines that are no longer needed

// Wait for any pending machines

// Create any machines that need creating

```

At the moment, if there is a pending Machine, this blocks any further creation right? This would mean, when surge is not 1, we can't actually surge over 1 if a Machine is pending.

Likewise, if there are two Machines in an index that both need an update, but one is pending, we wouldn't be able to continue. Eg, what happens if I update my CPMS with an invalid spec, new machine is pending forever, then I update the spec to fix it. I would expect the CPMS to delete the pending Machine and create a new Machine from the new spec. I don't think the logic right now allows for that

File: pkg/controllers/controlplanemachineset/updates.go (hunk 15 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go:444]: That's a good idea 👍

File: pkg/controllers/controlplanemachineset/updates.go (hunk 12 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go:361]: Yes these names where mainly a placeholder as I wanted to chat with others to see what they thought would've been good names for them (naming is hard). I like your suggestions for this and the other functions, I'll adopt them 👍

File: pkg/controllers/controlplanemachineset/updates.go (hunk 13 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go:342]: See https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/56#discussion_r923403427

File: pkg/controllers/controlplanemachineset/updates.go (hunk 14 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go:387]: See https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/56#discussion_r923403427

File: pkg/controllers/controlplanemachineset/updates.go (hunk 15 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go:444]: See https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/56#discussion_r923403427

File: pkg/controllers/controlplanemachineset/updates.go [hunk 17]
```diff
@@ -55,6 +56,10 @@ const (
 	// the current set of Machines.
 	noUpdatesRequired = "No updates required"
 
+	// noRoomForReplacement is a log message used to inform the user that no room for creating
+	// a replacement Machine is left as maxSurge has been reached.
+	noRoomForReplacement = "No room for creating a replacement Machine, maxSurge reached"
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:61]: I think capacity may be a better word to use here. Perhaps we want to say something like `Insufficient capacity for expansion. Will not create replacement Machine.`?

File: pkg/controllers/controlplanemachineset/updates.go [hunk 18]
```diff
@@ -152,3 +195,235 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineRollingUpdate(ctx con
...
+		if machinesInfo[i].Ready {
+			result = append(result, machinesInfo[i])
+		}
+	}
+
+	return result
+}
+
+// sortMachineInfosByIndex returns a list numerically sorted by index, of each index' MachineInfos.
+func sortMachineInfosByIndex(indexedMachineInfos map[int32][]machineproviders.MachineInfo) [][]machineproviders.MachineInfo {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:257]: This function would be good to have some individual unit tests if we don't already have them

File: pkg/controllers/controlplanemachineset/updates.go [hunk 19]
```diff
@@ -137,6 +142,44 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineUpdates(ctx context.C
...
+	// Devise the existing surge and keep track of the current surge count.
+	// No check for early stoppage is done here,
+	// as deletions can continue even if the maxSurge has been already reached.
+	surgeCount := deviseExistingSurge(cpms, sortedIndexedMs)
+
+	// Delete any machines that have successfully been replaced.
+	// This comes first so that we make room to create additional Machines later,
+	// otherwise a failure to update (perhaps because of surge capacity) may prevent deletes,
+	// in turn creating a deadlock scenario.
+	if done, result, err := r.deleteReplacedMachines(ctx, logger, sortedIndexedMs, machineProvider); err != nil || done {
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates.go:166]: Would like to discuss why we return early here if we delete a Machine, not sure that's necessary is it?

**Review by JoelSpeed — COMMENTED** [review]
Something we additionally need to consider here is what happens when you update the spec mid update, what happens if an index contains two machines, both in need of an update 🤔

File: pkg/controllers/controlplanemachineset/updates_test.go [hunk 20]
```diff
@@ -1274,3 +1388,54 @@ var _ = Describe("reconcileMachineUpdates", func() {
...
+var _ = Describe("utils tests", func() {
+	machineGVR := machinev1beta1.GroupVersion.WithResource("machines")
+	nodeGVR := corev1.SchemeGroupVersion.WithResource("nodes")
+	updatedMachineBuilder := resourcebuilder.MachineInfo().
+		WithMachineGVR(machineGVR).
+		WithNodeGVR(nodeGVR).
+		WithReady(true).
+		WithNeedsUpdate(false)
+
+	DescribeTable("should convert an indexed map of MachineInfo into a slice sorted by index",
```
[the remark this answers is not in the document]
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates_test.go:1401]: Shuffled around in https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/56/commits/f222e14e0d54acf9b79395d9ea94a35affe0b23a

File: pkg/controllers/controlplanemachineset/updates.go (hunk 17 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go:61]: We verbally agreed on `Insufficient capacity for expansion, maxSurge is currently being reached. Cannot create a replacement Machine at this time.`

File: pkg/controllers/controlplanemachineset/updates.go [hunk 21]
```diff
@@ -137,6 +142,44 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachineUpdates(ctx context.C
...
+		return result, err
+	}
+
+	// Wait for Pending Machine Replacements to become ready, if any exist.
+	if done, result, err := r.waitForPendingMachines(ctx, logger, sortedIndexedMs, machineProvider); err != nil || done {
+		return result, err
+	}
```
[the remark this answers is not in the document]
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go]: We talked about this offline, and carried on with the refactoring which allows multiple index and multiple action per index, in each reconcile loop.
See: https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/56/commits/fc2d4fe1ead847c53876127942b270cea12e128f and: https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/56/commits/574384e49043514e15f31614ab07bbc390393b72

File: pkg/controllers/controlplanemachineset/updates.go (hunk 19 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go]: See: https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/56#discussion_r926855940

File: pkg/controllers/controlplanemachineset/updates.go (hunk 16 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go]: See: https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/56#discussion_r926855940

File: pkg/controllers/controlplanemachineset/updates_test.go [hunk 22]
```diff
@@ -497,45 +545,320 @@ var _ = Describe("reconcileMachineUpdates", func() {
...
+					},
+					1: {
+						updatedMachineBuilder.WithIndex(1).WithMachineName("machine-1").WithNodeName("node-1").WithNeedsUpdate(true).Build(),
+						updatedMachineBuilder.WithIndex(1).WithMachineName("machine-replacement-1").WithNodeName("node-replacement-1").Build(),
+					},
+					2: {},
+				},
+				setupMock: func() {
+					mockMachineProvider.EXPECT().CreateMachine(gomock.Any(), gomock.Any(), gomock.Any()).Times(0)
+					mockMachineProvider.EXPECT().DeleteMachine(gomock.Any(), gomock.Any(), gomock.Any()).Times(1)
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates_test.go:686]: Which machine is being deleted?

File: pkg/controllers/controlplanemachineset/updates_test.go [hunk 23]
```diff
@@ -497,45 +545,320 @@ var _ = Describe("reconcileMachineUpdates", func() {
...
+								"index", 2,
+								"namespace", namespaceName,
+								"name", "<Unknown>",
+							},
+							Message: noCapacityForExpansion,
+						},
+					}
+				},
+			}),
+			Entry("with an extra index, hitting maxSurge", rollingUpdateTableInput{
```
Comment (JoelSpeed) [inline pkg/controllers/controlplanemachineset/updates_test.go:723]: If there's an extra index, I would the CPMS to remove the additional Machine

**Review by JoelSpeed — COMMENTED** [review]
This is looking nearly ready to go I think, couple of qs on the tests

File: pkg/controllers/controlplanemachineset/updates_test.go (hunk 23 above)
Comment (JoelSpeed) [reply pkg/controllers/controlplanemachineset/updates_test.go:723]: Discussed with Dam synchronously and we believe this case is valid, we will handle additional indexes in an earlier part of the code

File: pkg/controllers/controlplanemachineset/updates_test.go (hunk 22 above)
Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates_test.go]: Pinned in https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/56/commits/dd21fa703f3265703dd24087ce003ff995f640de
